File: backend/app/schemas/mvp.py
from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel, Field, HttpUrl, conint, constr


class UploadPresignRequest(BaseModel):
    workspace_id: conint(ge=1)
    filename: constr(min_length=1)
    mime_type: constr(min_length=1)
    size_bytes: conint(gt=0)


class UploadPresignResponse(BaseModel):
    upload_id: int
    storage_key: str
    url: str
    headers: Dict[str, str]
    max_mb: int


class UploadCompleteResponse(BaseModel):
    upload: Dict[str, Optional[str]]
    validate: Dict[str, Optional[str]]
    ingest: Dict[str, Optional[str]]


class WorkflowStatusResponse(BaseModel):
    workflow_id: str = Field(alias="id")
    step: str
    state: str
    progress: Optional[int] = None
    error: Optional[str] = None


class WorkflowTriggerRequest(BaseModel):
    dataset_id: str


class DatasetResponse(BaseModel):
    schema: List[Dict[str, str]]
    preview: List[Dict[str, str]]
    row_count: Optional[int]
    table_name: Optional[str]


class DashboardGenerateRequest(BaseModel):
    dataset_id: str


class ChartSpec(BaseModel):
    id: str
    title: str
    spec: Dict[str, object]


class DashboardGenerateResponse(BaseModel):
    charts: List[ChartSpec]
    layout: Dict[str, object]


class AIAskRequest(BaseModel):
    workspace_id: int
    dataset_id: str
    question: constr(min_length=2)


class AIAskResponse(BaseModel):
    answer: str
    usage_ms: int


class NarrativeBlock(BaseModel):
    heading: Optional[str] = None
    body: Optional[str] = None


class Branding(BaseModel):
    logo_url: Optional[HttpUrl] = None
    primary_color: Optional[str] = None


class ReportGenerateRequest(BaseModel):
    dataset_id: str
    selected_chart_ids: List[str] = Field(default_factory=list)
    narrative_blocks: List[NarrativeBlock] = Field(default_factory=list)
    branding: Optional[Branding] = None


class ReportGenerateResponse(BaseModel):
    report_id: str
    download_url: str

File: frontend/src/api/mvp.ts
import { api } from "@/lib/api";

export type UploadPresignPayload = {
  workspaceId: number;
  filename: string;
  mimeType: string;
  sizeBytes: number;
};

export type UploadPresignResponse = {
  upload_id: number;
  storage_key: string;
  url: string;
  headers: Record<string, string>;
  max_mb: number;
};

export type WorkflowStepState = {
  step: string;
  state: string;
  error?: string;
  id?: string;
  dataset_id?: string | null;
};

export type WorkflowSummary = {
  upload: WorkflowStepState;
  validate: WorkflowStepState;
  ingest: WorkflowStepState;
};

export type WorkflowStatus = {
  id: string;
  step: string;
  state: string;
  progress?: number | null;
  error?: string | null;
};

export type DatasetResponse = {
  schema: Array<Record<string, unknown>>;
  preview: Array<Record<string, unknown>>;
  row_count: number | null;
  table_name: string | null;
};

export type DashboardChart = {
  id: string;
  title: string;
  spec: Record<string, unknown>;
};

export type DashboardGenerateResponse = {
  charts: DashboardChart[];
  layout: Record<string, unknown>;
};

export type AIAskResponse = {
  answer: string;
  usage_ms: number;
};

export type ReportGenerateResponse = {
  report_id: string;
  download_url: string;
};

export async function presignUpload(payload: UploadPresignPayload) {
  const body = {
    workspace_id: payload.workspaceId,
    filename: payload.filename,
    mime_type: payload.mimeType,
    size_bytes: payload.sizeBytes,
  };
  const { data } = await api.post<UploadPresignResponse>("/uploads/presign", body);
  return data;
}

export async function uploadToPresignedUrl(url: string, file: File, headers: Record<string, string>) {
  const response = await fetch(url, {
    method: "PUT",
    headers,
    body: file,
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Upload failed: ${response.status} ${text}`);
  }
}

export async function completeUpload(uploadId: number) {
  const { data } = await api.post<WorkflowSummary>(`/uploads/${uploadId}/complete`);
  return data;
}

export async function getUploadWorkflow(uploadId: number) {
  const { data } = await api.get<WorkflowSummary>(`/uploads/${uploadId}/workflow`);
  return data;
}

export async function getWorkflowStatus(workflowId: string) {
  const { data } = await api.get<WorkflowStatus>(`/workflows/${workflowId}/status`);
  return data;
}

export async function triggerIngest(datasetId: string) {
  const { data } = await api.post<WorkflowStatus>(`/workflows/${datasetId}/ingest`);
  return data;
}

export async function getDataset(datasetId: string) {
  const { data } = await api.get<DatasetResponse>(`/datasets/${datasetId}`);
  return data;
}

export async function generateDashboards(datasetId: string) {
  const { data } = await api.post<DashboardGenerateResponse>("/dashboards/generate", {
    dataset_id: datasetId,
  });
  return data;
}

export async function askAI(workspaceId: number, datasetId: string, question: string) {
  const { data } = await api.post<AIAskResponse>("/ai/ask", {
    workspace_id: workspaceId,
    dataset_id: datasetId,
    question,
  });
  return data;
}

export async function generateReport(payload: {
  datasetId: string;
  selectedChartIds: string[];
  narrativeBlocks: { heading?: string; body?: string }[];
  branding?: { logo_url?: string; primary_color?: string };
}) {
  const { data } = await api.post<ReportGenerateResponse>("/reports/generate", {
    dataset_id: payload.datasetId,
    selected_chart_ids: payload.selectedChartIds,
    narrative_blocks: payload.narrativeBlocks,
    branding: payload.branding,
  });
  return data;
}
